src/nvim/terminal.c

```diff
@@ -169,6 +170,50 @@ static VTermScreenCallbacks vterm_screen_callbacks = {
   .sb_popline  = term_sb_pop,
 };
 
+static void emit_osc_event(void **argv)
+{
+  int command = (int)(ptrdiff_t)argv[0];
+  char *payload = argv[1];
+  int payload_length = (int)(ptrdiff_t)argv[2];
+
+  save_v_event_T save_v_event;
+  dict_T *dict = get_v_event(&save_v_event);
+
+  typval_T cmd;
+  cmd.v_type = VAR_NUMBER;
+  cmd.vval.v_number = command;
+  tv_dict_add_nr(dict, S_LEN("command"), cmd.vval.v_number);
+
+  typval_T str;
+  str.v_type = VAR_STRING;
+  str.vval.v_string = payload;
+  tv_dict_add_str_len(dict, S_LEN("payload"), str.vval.v_string, payload_length);
+
+  tv_dict_set_keys_readonly(dict);
+
+  apply_autocmds(EVENT_TERMOSC, NULL, NULL, false, curbuf);
+
+  restore_v_event(dict, &save_v_event);
+  xfree(payload);
+}
+
+static int on_osc(int command, VTermStringFragment frag, void *user)
+{
+  char *str = xmemdupz(frag.str, frag.len);
+  multiqueue_put(main_loop.events, emit_osc_event, 3, command, str, frag.len);
```

Comment:
Instead of passing the command as a separate field from the payload, I suggest putting the command back into the sequence so that TermRequest handlers can parse the OSC sequence directly (this is how TermResponse works).

So e.g. instead of checking `v:event.command` and `v:event.payload`, have only `v:event.termrequest` (or make a `v:termrequest` analogous to `v:termresponse`? Or maybe `v:termresponse` should be moved into `v:event`?) which contains the `\x1b]{command}` prefix.

```suggestion
  StringBuilder request = KV_INITIAL_VALUE;
  kv_printf(request, "\x1b]%d%s", command, frag.str);
  multiqueue_put(main_loop.events, emit_osc_event, 1, request.items, request.size);
```